### PR TITLE
User database implementation

### LINKED ISSUE #72: Data model: User DB implementation
Define the database schema of the user model and the table class for it.

### PR DESCRIPTION
**Pull request by Plouc314**
* Create `Table` abstract class, refactore `AvailabilityTable`. Avoid some boilerplate for simple operations.
* Create `UserSchema`, `UserTable` classes
* `assignedFlights` is not handled until #85 is implemented
* `roleTypes` is not handled as it is a private attribute at the moment
* The test setup will be simplified in (#106)
* The `SystemClock.sleep` is not an ideal solution, we can improve on it later

Close #72

### CODE REVIEW

File: app/src/main/java/ch/epfl/skysync/models/calendar/CalendarModel.kt [hunk 1]
```diff
@@ -109,4 +110,22 @@ abstract class CalendarModel<T : CalendarViewable> {
   fun getByDate(date: LocalDate, timeSlot: TimeSlot): T? {
     return cells.firstOrNull { it.date == date && it.timeSlot == timeSlot }
   }
+
+  /** Return a set of the cells, sorted by date and time slot */
+  private fun getSortedSetOfCells(): SortedSet<T> {
```
Comment (RAFInery) [inline app/src/main/java/ch/epfl/skysync/models/calendar/CalendarModel.kt:115]: we might consider at some point to directly implement and maintain the underlying list in `CalendarModel` as a sorted structure

File: app/src/main/java/ch/epfl/skysync/database/Table.kt [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  /**
+   * Delete the table
+   *
+   * This is only used for testing, as such it is only supported if using the emulator.
+   *
+   * @param onError Callback called when an error occurs
+   */
+  open fun deleteTable(onError: (Exception) -> Unit) {
```
Comment (RAFInery) [inline app/src/main/java/ch/epfl/skysync/database/Table.kt:70]: maybe define in a separate subclass for the emulator?

File: app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * This class allows for scheduling a callback function to be executed after a specified delay
+ * period. It provides functionality to decrement the delay period on each invocation of the `run()`
+ * method until the delay reaches 0, at which point the callback function is executed.
+ *
+ * @property delay The initial delay period before executing the callback, specified in units of
+ *   invocation cycles.
+ * @property callback The callback function to be executed after the delay period.
+ */
+class DelayedCallback(private var delay: Int, private val callback: () -> Unit) {
```
Comment (RAFInery) [inline app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt:15]: I didn't really understand the use of DelayedCallback. Why did you choose this approach and generally, how do you choose the appropriate delay?

**Review by RAFInery — APPROVED** [review]
Seems fine. DB interactions are nicely encapsulated and appear to cover already a vast amount of queries we'll need

File: app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt (hunk 3 above)
Comment (Plouc314) [reply app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt:15]: At the moment I use the callback approach to asynchronous functions, as from what I looked the "async/await" support of Kotlin is pretty poor, the delayed callback is used when launching multiple asynchronous functions "in parallel", as each function takes a callback, but we want to execute the "final" callback only once after all functions are done. You can find an example of utilization in UserTable.deleteAvailabilities`

Comment (RAFInery) [reply app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt:15]: got  it. Maybe update the description to something like:

_If n similar asynch. operations are executed (in a random order) but the call function should be called only once by the last of the n operations that finishes _ and change the signature to

`DelayedCallback(private var delay: Int, private val callback: () -> Unit)` => 
`ExecuteCallbackOnceByLastOperation(private var nParallelOperations: Int, private val callback: () -> Unit)`

Comment (Plouc314) [reply app/src/main/java/ch/epfl/skysync/database/DelayedCallback.kt]: Renamed it to `ParallelOperationsEndCallback`

**Pull request merged**